### DISCUSSION TITLE
Migrating to Synology NAS from MBP

### DISCUSSION THREAD

**Discussion by noiwontsmile** [philosowaffle/peloton-to-garmin#184] [Q&A]
I setup p2g using docker-compose on my MBP and it's running well and now I'm attempting to migrate it over to my Synology NAS.  First off I upgraded the docker-compose on the NAS to 1.29.2.  Then I copied the entire p2g folder (containing all the sub directories for p2g, *.yaml, *.json, etc...) from my MBP and dropped it into the /volume1/docker/ folder on the NAS.  From there I ssh'ed into the NAS, then switched to sudo, cd'ed into the p2g directory and ran docker-compose up.  

Thing we're looking good, meaning I saw the container get created and start but then it failed because it couldn't make the http call to Peloton.  Whole log is pasted at bottom.  I think I know enough to confidently say the container isn't able to make a connection to the internet but I don't know enough on how to resolve the issue.  Other tidbits of information: I've installed Portainer; the firewall isn't turned on the NAS;  I'm currently running pihole and plex containers on the same NAS with no issues; I did install vault warden (used this guide: https://www.wundertech.net/how-to-self-host-the-password-manager-bitwarden-on-a-synology-nas/) but never finished setting it up so the container isn't running but part of that install included a proxy.  

So with that brain dump, any thoughts on how I can get past this current error?  Thanks. 

---- Log below----
2021-10-30 10:15:48.006 -06:00 [WRN] App.PythonAndGuploadInstalledLocally setting is deprecated and will be removed in a future release. Please swith to using Garmin.UploadStrategy config.
2021-10-30 10:15:58.731 -06:00 [ERR] Http Call Failed. null 
2021-10-30 10:15:58.742 -06:00 [ERR] Failed to authenticate with Peloton.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Common.FlurlConfiguration.<>c__DisplayClass1_0.<<Configure>b__1>d.MoveNext() in /build/src/Common/FlurlConfiguration.cs:line 35
--- End of stack trace from previous location ---
   at Flurl.Http.FlurlRequest.SendAsync(HttpMethod verb, HttpContent content, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Flurl.Http.ResponseExtensions.ReceiveJson[T](Task`1 response)
   at Peloton.ApiClient.InitAuthAsync(String overrideUserAgent) in /build/src/Peloton/ApiClient.cs:line 38
2021-10-30 10:15:58.781 -06:00 [INF] No download directory found. Nothing to do. ./working/downloaded
2021-10-30 10:15:58.783 -06:00 [INF] No upload directory found. Nothing to do.
2021-10-30 10:15:58.784 -06:00 [INF] Done.

**Comment by philosowaffle** [comment]
Frankly, I'm stumped on this one. Might need to sleep on it and see if I can come up with any theories.

My only current theory would be that your configuration file is not actually getting mounted into the container resulting in an authentication error with Peloton and the `Http Call Failed` message could be a red herring.

**Reply by noiwontsmile** [reply]
So thinking about what you said "...configuration file is not actually getting mounted into the container..." I thought I'd export a yaml file using docker-autocompose (https://github.com/Red5d/docker-autocompose) to see what would fall out.  I plan to use docker-autocompose to migrate all (both) of my other containers to docker-compose in the future. 

Below is the output from docker-autocompose.  _Please point out if there is something sensitive in the message below that I shouldn't have posted.  I know it will be too late but I prefer to learn from my mistakes._
----docker-autocompose output----
version: "3"
services:
  p2g2:
    container_name: p2g2
    cpu_shares: 50
    entrypoint:
      - ./PelotonToGarminConsole
    environment:
      - PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
      - ASPNETCORE_URLS=http://+:80
      - DOTNET_RUNNING_IN_CONTAINER=true
      - DOTNET_VERSION=5.0.10
      - ASPNET_VERSION=5.0.10
      - PYTHONUNBUFFERED=1
      - TZ=America/MyHome
    hostname: 95628d6db2f6
    image: philosowaffle/peloton-to-garmin
    ipc: private
    labels:
      com.docker.compose.config-hash: 503125b5a0eda54fc16375cc0298da69f18e36671e62330ebd12c7c6bbf466f6
      com.docker.compose.container-number: 1
      com.docker.compose.oneoff: False
      com.docker.compose.project: p2g_take2
      com.docker.compose.project.config_files: docker-compose.yaml
      com.docker.compose.project.working_dir: /volume1/docker/p2g_take2
      com.docker.compose.service: p2g
      com.docker.compose.version: 1.29.2
    logging:
      driver: db
      options: {}
    networks:
      p2g_take2_default:
        aliases:
          - 95628d6db2f6
          - p2g
    volumes:
      - /volume1/docker/p2g_take2/working:/app/working:rw
      - /volume1/docker/p2g_take2/device_info.xml:/opt/app/device_info.xml:rw
      - /volume1/docker/p2g_take2/configuration.local.json:/app/configuration.local.json:rw
      - /volume1/docker/p2g_take2/author_info.xml:/opt/app/author_info.xml:rw
      - /volume1/docker/p2g_take2/output:/app/output:rw
    working_dir: /app
networks:
  p2g_take2_default:
    external: true

**Comment by philosowaffle** [comment]
Just a quick sanity check the below path exists and has your expected configuration?

```
/volume1/docker/p2g_take2/configuration.local.json
```

Second sanity check, verify the permissions on the volumes mounted to the container. Since you copied them over its possible they were created by the Root user and the docker container would not be running as Root and so therefore not able to access them.

I know none of this has anything to do with the HTTP error, but I've still got no ideas on that front.  My theory is still that it might be a config or permissions issue, and the HTTP error is a red herring... but its just a guess.

**Reply by noiwontsmile** [reply]
Yep, that is the path where the configuration.local.json file is.  I reviewed all the permissions and I believe everything is correct and aligned all through the containers files and subdirectories.  Just to be sure I also renamed the configuration.local.json file and an error was thrown saying it couldn't find the file.

I went on and created another container folder through the command line on the NAS and touched all the file and copied in all the necessary data.  Same error.  

I also installed docker on a different computer and used the same "source" p2g container and contents from the original MBP and everything worked fine.

**Reply by philosowaffle** [reply]
Hmmm ok so its definitely something specific to the deployment on the synology NAS since it seems to run fine on a different computer.  

Unfortunately, I haven't ever run anything with docker on my synology NAS so I'm not versed on any differences there might be. I know Synology OS is different than your standard Linux or Windows OS.

I did some googling and found [this forum post](https://community.synology.com/enu/forum/11/post/142221) that sounds similar to your issue. The root cause seems to have been something related to pi-hole and DNS resolution.

Edit: seems the "solution" blames synology firewall, but you already disabled firewall, so not a solution for you.
